### PR TITLE
MM-37874: Add zoom + pan function to image preview

### PR DESCRIPTION
**Pull request by shamboozles**
#### Summary

Adds zoom and pan functionality to image preview along with a zoom control toolbar, similar to the one in PDF.js

Also, the toolbar text (Automatic, Fit width, Fit height) is international but only has English, French, Spanish, Italian and German translations.

#### Ticket Link

Related issue: https://mattermost.atlassian.net/browse/MM-37874

#### Screenshots

|  before  |  after  |
|----|----|
| Screenshot 2022-08-11 at 09 40 21 | https://user-images.githubusercontent.com/108882245/184133772-2ad24a59-4ef2-49bb-8ca4-30acf7589d12.mp4 |

#### Release Note

```release-note

* Added image preview control support

```

### CODE REVIEW

File: components/file_preview_modal/image_preview.jsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.jsx:172]: 👍🏼

File: components/file_preview_modal/image_preview.jsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.jsx:123]:
```suggestion

        if (zoom <= maxCanvasZoom) {
            return {xPos: 0, yPos: 0};
        }

        return {
            xPos: horizontal ? clamp(x, w, -w) : 0,
            yPos: vertical ? clamp(y, h, -h) : 0
        };
```

File: components/file_preview_modal/image_preview.jsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.jsx:142]:
```suggestion
        if (!dragging) {
            return;
        }
        
        const {x, y} = touch.current;
        const {clientX, clientY} = event;
        const {xPos, yPos} = clampOffset(offset.x + (x - clientX), offset.y + (y - clientY));
        setOffset({x: xPos, y: yPos});
        touch.current = {x: clientX, y: clientY};
```

File: components/file_preview_modal/image_preview.jsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.jsx:168]:
```suggestion
    const handleMouseEnter = () => setDragging(isMouseDown.current);
```

File: components/file_preview_modal/image_preview.jsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.jsx:161]:
```suggestion
    const handleMouseLeave = () => setDragging(false);
```

File: components/file_preview_modal/image_preview.jsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.jsx:180]: interesting use of `useState` to force a re-render! :D

File: i18n/es.json
Comment (michelengelen) [inline i18n/es.json:3206]: please only change the `en.json` file for translations ... all others are automatically generated from weblate

**Review by michelengelen — CHANGES_REQUESTED** [review]
I am sorry about all the comments and suggestions. It generally looks good. I particularly like your comments on each of the steps. Nicely done!

**Comment by shamboozles** [comment]
> I am sorry about all the comments and suggestions. It generally looks good. I particularly like your comments on each of the steps. Nicely done!

No worries, I appreciate them, it's what helps the code to improve :)
Thanks for all the comments, I'll see to them soon!

File: components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.tsx
Comment (shamboozles) [reply components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.tsx:80]: I guess we could but then this happens:
Screenshot 2022-08-16 at 09 16 56

I could fix it but is it worth it?

Comment (michelengelen) [reply components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.tsx:80]: it depends ... I would prefer a tidy structure when it is just a minor adjustment

File: components/file_preview_modal/image_preview.jsx
Comment (shamboozles) [reply components/file_preview_modal/image_preview.jsx:20]: I actually ended up removing them as I could export the variables `zoom` and `minZoom` directly from the `image_preview` component. Is this ok?

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (shamboozles) [reply components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:48]: I actually don't think I should do that as the values are automatically converted to strings in HTML. But I can indeed remove the `toString`.

File: components/file_preview_modal/file_preview_modal.scss
Comment (shamboozles) [reply components/file_preview_modal/file_preview_modal.scss:25]: Yeah, I actually dont remember why I deleted that line 😅

File: components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.tsx
Comment (shamboozles) [reply components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.tsx:80]: I just remembered that now but the reason I separated them like that is because they are different divs and are stacked, it made sense for me te separate them. Please tell me if that isn't the correct way to do it and I will rectify. Alternatively, I could do that:

```
        <div>
            <div className='file-preview-modal-header'>
                {actionProps.isMobileView && actions}
                {!actionProps.isMobileView &&
                <FilePreviewModalInfo
                    showFileName={true}
                    post={post}
                    filename={actionProps.filename}
                />
                }
                {mainActions}
                {!actionProps.isMobileView && actions}
            </div>
            {imageControls}
        </div>
```

File: components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.test.tsx
Comment (shamboozles) [reply components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.test.tsx:27]: To be honest, I never wrote tests before.. I tried something, let me know if that is what you expected

File: components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.tsx
Comment (michelengelen) [reply components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.tsx:80]: that's not really better ... I generally like to keep the HTML clutter as minimal as possible to not create overly complex and nested structures. Let's just keep the Fragment for now and when we revisit this component we can surely improve on it then. For now it's fine and I don't want to block the PR with nitpicky stuff like this!

**Comment by michelengelen** [comment]
@shamboozles let me know when this is ready for another round of reviews! 👍🏼

**Comment by shamboozles** [comment]
I'd say its ready :D

File: components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.test.tsx
Comment (michelengelen) [inline components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.test.tsx:92]: well done ... you could also add observers to the zoomhandler and emulate a click on the buttons to see if they are called with the right values. see `components/setting_item_max.test.jsx` as an example.

Also wouldn't adding E2E tests here a good option @jgilliam17 ?

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (michelengelen) [inline components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:31]:
```suggestion
    const plusSign = (
        <PlusIcon
            size={16}
            color='currentColor'
        />
    );
    const minusSign = (
        <MinusIcon
            size={16}
            color='currentColor'
        />
    );
```

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (michelengelen) [inline components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:93]: what about options `'W'` and `'H'`?

File: components/file_preview_modal/image_preview.tsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.tsx:103]:
```suggestion

        return {
```

**Review by michelengelen — APPROVED** [review]
Great work ... I will unblock and if you have time I would appreciate if you could add some more tests and possibly E2E tests for this feature. Lets wait for @jgilliam17 on that, though!

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (shamboozles) [reply components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:93]: Oh 🤦 
I had temporarily put this there to avoid lint errors while trying to find a better solution

File: components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.test.tsx
Comment (shamboozles) [reply components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.test.tsx:92]: Yep, I'll do that today

**Comment by shamboozles** [comment]
Thanks for the review! By the way, I'm terribly sorry, I was a little tired yesterday and some text that shouldn't have been there made it's way into the commit 😅

File: components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.test.tsx
Comment (shamboozles) [reply components/file_preview_modal/file_preview_modal_header/file_preview_modal_header.test.tsx:92]: I've finally had some time to get to it, sorry about that. They're done now.. Please tell me if there's anything that I should add

**Comment by shamboozles** [comment]
@michelengelen  I've finally had some time to get to the tests, sorry about that. They're done now.. Please tell me if there's anything that I should add

File: package-lock.json
Comment (M-ZubairAhmed) [inline package-lock.json:5376]: please revert this file

File: components/file_preview_modal/image_preview.tsx
Comment (M-ZubairAhmed) [inline components/file_preview_modal/image_preview.tsx:210]: please use classname library here to construct class name

File: components/file_preview_modal/image_preview.tsx
Comment (M-ZubairAhmed) [inline components/file_preview_modal/image_preview.tsx:9]: css import can be at the bottom of the import order, since its relative import

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (michelengelen) [inline components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:40]: these can just go into the return value. No need to store them in variables upfront

File: components/file_preview_modal/image_preview.tsx
Comment (M-ZubairAhmed) [inline components/file_preview_modal/image_preview.tsx:166]: is this eslint error or warning? if warning then we dont have to include the disable comment

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (M-ZubairAhmed) [inline components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:5]: nit : mm dependencies are imported after 3rd party packages

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
Looks code wise good overall, few nit picks

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (michelengelen) [inline components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:13]: Sry I did not catch this earlier: It is generally not a good idea to have variables declared outside of a component. This is fine for constants, but whenever you are mutating the values it is better to use a `ref`

File: components/file_preview_modal/image_preview.tsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.tsx:22]: same thing applies here: please use a `ref` when you need variables to be mutated and not in state

File: components/file_preview_modal/image_preview.tsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.tsx:80]:
```suggestion
    const containerRef = useRef<HTMLDivElement>(null);
```

File: components/file_preview_modal/image_preview.tsx
Comment (michelengelen) [inline components/file_preview_modal/image_preview.tsx:13]: nit: 

```suggestion
import {FileInfo} from '@mattermost/types/files';

import {ZoomValue} from './file_preview_modal_image_controls/file_preview_modal_image_controls';
import {LinkInfo} from './types';

import './image_preview.scss';
```

**Review by michelengelen — APPROVED** [review]
just a few comments from my end: Code looks 🪨 -solid! Thanks a lot!

Tests are also fine .. good work! 🚀

File: components/file_preview_modal/image_preview.tsx
Comment (shamboozles) [reply components/file_preview_modal/image_preview.tsx:23]: Would I still be able to export them?

Comment (shamboozles) [reply components/file_preview_modal/image_preview.tsx:23]: @michelengelen I can't figure out a way to use ref while still exporting them..

Comment (shamboozles) [reply components/file_preview_modal/image_preview.tsx]: I tried something but I'm not sure it's ideal

**Review by M-ZubairAhmed — APPROVED** [review]

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
This is awesome! Really great work @shamboozles. A few minor details:

1. Can we adjust the color of the bar these controls sit on? It should be: #222224
2. Instead of SVGs for the +/i icons, can we use Compass Icons? @michelengelen can probably help with this. And then we should adjust the size of the icons to use 18px font-size.
3. The +/- buttons should be 32x32 clickable areas and can we add the following interactive states to the `+` and `-` buttons?

image

5. The style of our dropdowns has changed since this design work was done. Would it be possible to update that?

image

I've put these together in the figma file here: https://www.figma.com/file/i86bHWcB68HMiaJNoVxmsq/Image-Preview-Modal?node-id=2960%3A45246

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (michelengelen) [inline components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:8]: @matthewbirtch the compass icon components are already being used by @shamboozles

**Comment by shamboozles** [comment]
I pushed the style changes. Sadly, I will not work on this after this day. My colleague @antonbuks who has been following the changes is willing to take over once I am gone, if any other modifications are needed after my leave

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (matthewbirtch) [reply components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:8]: Ok, interesting @michelengelen. Do they always output as SVGs then? I thought they outputted as `i` tags with classes.

This is what the button tag HTML is showing for me:

image

**Review by matthewbirtch — APPROVED** [review]
Thanks @shamboozles the style changes look great!

**Comment by shamboozles** [comment]
You're welcome, my pleasure!

File: components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx
Comment (michelengelen) [reply components/file_preview_modal/file_preview_modal_image_controls/file_preview_modal_image_controls.tsx:8]: @matthewbirtch yes, it is always outputting `svg` elements. This is due to a better ability to position the icon, since it is not controlled by font properties in css, but a block positioned element instead. 

Using `i` as an element to display icons is a remnant of old times and the usage is semantically incorrect, since the intention of the `i` tag is a different one. For reference: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/i

**Comment by M-ZubairAhmed** [comment]
@shamboozles can you please resolve the conflicts so we can forward for QA testing

**Comment by antonbuks** [comment]
@shamboozles asked me to help him finish this as he is not able to work on it atm but I don't have the push permissions for his fork I just realized, I will try to see if he can add me, otherwise I'll have to merge and reopen from my fork if he's not available.

**Comment by M-ZubairAhmed** [comment]
@antonbuks how can we help you to get this to finish line

**Comment by antonbuks** [comment]
Haven't been able to get the rights for this fork so I'll have to reopen from my own, I'll cross post here when I have a new merge request open. I'll try to get it wrapped up this week just a little busy atm.

**Comment by antonbuks** [comment]
New merge request open to continue this here https://github.com/mattermost/mattermost-webapp/pull/11152

**Comment by michelengelen** [comment]
closing this in favor of #11152

**Comment by vetash** [comment]
That's look nice, but there is also need to add rotate functions.

**Comment by michelengelen** [comment]
> That's look nice, but there is also need to add rotate functions.

@vetash What exactly do you need with this? Are you missing a feature to rotate a picture in here? If so please feel free to add a feature request [here](https://community.mattermost.com/core/channels/feature-ideas). It certainly sounds like a good addition 🚀

**Pull request closed without merging**
